### PR TITLE
fix: Report text trimmed by MaxLines

### LINKED ISSUE #21043: TextBlock text trimming is not working properly when it is inside the property TabViewItem.Header
### Current behavior 🐛

When having a `TextBlock` inside the `TabViewItem.Header` property, the property `TextBlock.IsTextTrimmed` is not set as `true` and the `TextBlock_IsTextTrimmedChanged` event handler is not being raised.

### Expected behavior 🎯

Have the property `IsTextTrimmed` with `true` as value and execute the event handler.

### How to reproduce it (as minimally and precisely as possible) 🔬

1. Open the attached solution and compile it.
2. Put a breakpoint inside the event handler in the MainPage.xaml code-behind.
3. Execute it.
4. See that does not go through the event handler except in WinUI.

### Workaround 🛠️

No workaround found.

### Renderer 🎨

- [x] Skia
- [x] Native

### Affected platforms 📱💻🖥️

Desktop (Windows), WebAssembly, Android

### Uno.Sdk version (and other relevant versions) 📦

Uno.Sdk version: 6.0.146
Uno.WinUI version: 6.0.797

### IDE version 🧑‍💻

Visual Studio 2022 17.14.8

### Anything else we need to know? 💬

Repro: [TabViewItemTrimmingIssue.zip](https://github.com/user-attachments/files/21227667/TabViewItemTrimmingIssue.zip)

Urgency: **High**
Impact: **Low**

### PR DESCRIPTION
**Pull request by erikvilima**
**GitHub Issue:** closes #21043

## PR Type:
- 🐞 Bugfix

## PR Checklist ✅

Please check if your PR fulfills the following requirements:

- [ ] 📝 Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.
- [ ] 🧪 Added [Runtime tests, UI tests, or a manual test sample](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] 📚 Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] 🖼️ Validated PR `Screenshots Compare Test Run` results.
- [ ] ❗ Contains **NO** breaking changes

### CODE REVIEW

File: src/Uno.UI/UI/Xaml/Documents/InlineCollection.skia.cs [hunk 1]
```diff
@@ -177,6 +178,7 @@ internal Size Measure(Size availableSize, float defaultLineHeight)
 
 						if (maxLines > 0 && _renderLines.Count == maxLines)
 						{
+							_isTrimmedByMaxLines = run.Segments.IndexOf(segment) != run.Segments.Count - 1;
```
Comment (Xiaoy312) [inline src/Uno.UI/UI/Xaml/Documents/InlineCollection.skia.cs]:
```suggestion
							_isTrimmedByMaxLines = run.Segments[^1] == segment
```

File: src/Uno.UI/UI/Xaml/Documents/InlineCollection.skia.cs [hunk 2]
```diff
@@ -1094,6 +1096,8 @@ private int AdjustIndexForSurrogatePairs(int index)
 
 		internal float AverageLineHeight => _renderLines.Count > 0 ? _renderLines.Average(r => r.Height) : _lastDefaultLineHeight;
 
+		internal bool IsTrimmedByMaxLines() => _isTrimmedByMaxLines;
```
Comment (Xiaoy312) [inline src/Uno.UI/UI/Xaml/Documents/InlineCollection.skia.cs]: no point making this method, this is pretty textbook side-effect free
```suggestion
		internal bool OverflowMaxLines { get; private set; }
```
and drop the field at the top

File: src/SamplesApp/UITests.Shared/Windows_UI_Xaml_Controls/TextBlockControl/TextBlock_IsTextTrimmedChanged.xaml [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+	<StackPanel Width="80">
+		<TextBlock IsTextTrimmedChanged="TextBlock_IsTextTrimmedChanged_Event"
+				   TextWrapping="Wrap"
+				   MaxLines="3"
+				   TextTrimming="CharacterEllipsis"
+				   x:Name="MyTextBlock"
+				   Text="Lorem ipsum dolor sit amet consectetur. Porta habitasse iaculis non egestas posuere in tincidunt. Risus nunc a  urna est nunc ac mattis eget. Faucibus ullamcorper ipsum faucibus in tristique lacus. Justo cursus arcu ut tristique." />
+	</StackPanel>
+</UserControl>
```
Comment (Xiaoy312) [inline src/SamplesApp/UITests.Shared/Windows_UI_Xaml_Controls/TextBlockControl/TextBlock_IsTextTrimmedChanged.xaml]: (there is no visual indication here... update the result to a textblock?)

but this could just a runtime test to avoid wasting qa time

Comment (Xiaoy312) [reply src/SamplesApp/UITests.Shared/Windows_UI_Xaml_Controls/TextBlockControl/TextBlock_IsTextTrimmedChanged.xaml]: that said, you should keep an ui test version
where the stack-panel width is bound to a slider so we can validate edge cases (the ones near the tipping point)
with a textbbock bound to IsTextTrimmed

**Review by Xiaoy312 — APPROVED** [review]

File: src/Uno.UI/UI/Xaml/Controls/TextBlock/TextBlock.Android.cs [hunk 4]
```diff
@@ -464,6 +454,14 @@ private Size UpdateLayout(ref LayoutBuilder layout, Size availableSize, bool exa
 				}
 			}
 
+			if (MaxLines > 0)
+			{
+				var newLayoutWithoutMaxLines = CreateLayoutBuilder(layout, availableSize, exactWidth, 0);
+				newLayoutWithoutMaxLines.Build();
+
+				_overflowMaxLines = newLayoutWithoutMaxLines.MeasuredSize.Height > layout.MeasuredSize.Height;
```
[the remark this answers is not in the document]
Comment (Xiaoy312) [reply src/Uno.UI/UI/Xaml/Controls/TextBlock/TextBlock.Android.cs:462]: yes, we should be caching those. but this is not the right way...
in LayoutBuilder' UpdateLayout methods we are already doing the overflow capping recalculation (in the last if-block), just get that info out, so we dont have to perform that calculation again for nothing

[the remark this answers is not in the document]
Comment (erikvilima) [reply src/Uno.UI/UI/Xaml/Controls/TextBlock/TextBlock.Android.cs:462]: I noticed that `Layout.LineCount` always seems to have the same value as `MaxLines` ([code](https://github.com/unoplatform/uno/blob/14baa9174b1b3311d96c0ae50257ae70b20dc851/src/Uno.UI/UI/Xaml/Controls/TextBlock/TextBlock.Android.cs#L691)), which means that condition never becomes `true`.

It also looks like this behavior comes from how `Layout` is assigned by the Android code, so it might not be something we can easily change on our side.  

Do you happen to have any suggestions or thoughts on how we might approach this?

**Review by Xiaoy312 — APPROVED** [review]

File: src/Uno.UI/UI/Xaml/Controls/TextBlock/TextBlock.Android.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (Xiaoy312) [reply src/Uno.UI/UI/Xaml/Controls/TextBlock/TextBlock.Android.cs:462]: i noticed we are keep multiple copies layouts already, a +1 would be fine too.
i dont have anything better to suggest here, so lets take your current approach.

File: src/SamplesApp/UITests.Shared/Windows_UI_Xaml_Controls/TextBlockControl/TextBlock_IsTextTrimmed.xaml [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+					   Text="Lorem ipsum dolor sit amet consectetur. Porta habitasse iaculis non egestas posuere in tincidunt. Risus nunc a  urna est nunc ac mattis eget. Faucibus ullamcorper ipsum faucibus in tristique lacus. Justo cursus arcu ut tristique."
+					   TextWrapping="Wrap"
+					   MaxLines="{Binding ElementName=WidthSlider, Path=Value}"
+					   TextTrimming="CharacterEllipsis" />
+
+		</StackPanel>
+
+		<TextBlock x:Name="TrimmedStatus">
+			<Run Text="IsTextTrimmed: " />
+			<Run x:Name="IsTextTrimmed" Text="{Binding ElementName=LongText, Path=IsTextTrimmed}" />
```
Comment (erikvilima) [inline src/SamplesApp/UITests.Shared/Windows_UI_Xaml_Controls/TextBlockControl/TextBlock_IsTextTrimmed.xaml:34]: This isn't working of native iOS.

File: src/SamplesApp/SamplesApp.UITests/Windows_UI_Xaml_Controls/TextBlockTests/TextBlockTests.cs [hunk 6]
```diff
@@ -432,6 +432,32 @@ public void When_Padding_Is_Changed_Then_Cache_Is_Missed()
 			w3.Should().Be(w1);
 		}
 
+		[Test]
+		[AutoRetry]
+		public void When_Overflow_MaxLines_IsTextTrimmed_Changed()
```
Comment (erikvilima) [inline src/SamplesApp/SamplesApp.UITests/Windows_UI_Xaml_Controls/TextBlockTests/TextBlockTests.cs:437]: Investigate why this isn't working for Native WASM.
